infrastructure/src/main/java/org/corfudb/infrastructure/BatchWriter.java

```diff
@@ -164,18 +173,95 @@ public void reset(@Nonnull long epoch) {
         }
     }
 
+    /**
+     * Request log tail, i.e., max address for this stream.
+     *
+     * @param epoch epoch for this request.
+     * @return address of the log's tail.
+     */
+    public long queryLogTail(long epoch) {
+        try {
+            CompletableFuture<Long> cf = new CompletableFuture<>();
+            operationsQueue.add(new BatchWriterOperation(Type.LOG_TAIL_QUERY, Address.NON_EXIST,
+                    LogData.EMPTY, epoch, Collections.EMPTY_LIST, cf));
+            return cf.get();
```

Comment:
can you rebase on https://github.com/CorfuDB/CorfuDB/pull/1837 it cleans up this class quite a bit.